### DISCUSSION TITLE
Sample of Unity app driving NightDriver over WiFi

### DISCUSSION THREAD

**Discussion by davepl** [PlummersSoftwareLLC/NightDriverStrip#508] [Announcements]
Just a quick demo I did of a Unity surface being sent to the NIghtDriver Spectrum project over wifi.

https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/3698967/fce1a9ef-a959-466f-9f9f-70fe6ec13a40

**Comment by rbergen** [comment]
I'm just noting that if anyone wants to reproduce this in the Spectrum configuration, it's necessary to set the `INCOMING_WIFI_ENABLED` define to 1 in the `SPECTRUM` block in globals.h. That does mean that the on-device web server won't come up any more, which means it can just as well be disabled.

**Comment by mikejohnau** [comment]
So I can get the Unity to throw video at both the spectrum and mesmeriser builds that I have. But have noticed a couple of issues.

Spectrum build is the standard M5stcikplus with 3 x 16x16 matrix.

https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/26905608/63e317fe-e154-4341-b041-3102f56b74d8

Works ok..... Brightness is huge! I have turn the alpha in unity down to around 0.1 otherwise blindness, skin cancer and other light related issues arise. And the neighbours complain........ but the colours look ok but a it washed out

Is there a way to set the a brightness level in Unity hat actually works. 


Then I threw it at my mesmerizer build ( a hardwired wrover kit). Chnaged the resolution i Unity to 64x32 and the IP address. It worked....I get video and movement... but a pink pacman??

https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/26905608/495413af-9eed-420e-89b2-203629991e3c

**Reply by heidepiek** [reply]
Could someone do an explanation for dummies. Have downloaded unity hub 3.6.1. Downloaded unity 2021.3.32f1.
Downloaded NightDriverUnity . But can't get it to work. 
In unity I open project nightdriverunity, but then it stops for me. It can't find the ip address. The program itself gives it an ip address. Been trying to get this done for a few hours now, but throwing in the towel. 😟 

Harrie

**Reply by mikejohnau** [reply]
Hi Harrie,
The first thing that you need to do is ensure that
#define INCOMING_WIFI_ENABLED       1   // Accepting incoming color data and commands
is enabled in globals.h when you build.

After that you shoud just be able to put your devices IP address into Unity and hit play depending on the ESP32 board that you have.

I found with my config that I get better performance by increasing the amount of buffers in globals.h to around 40 as well.

You could also have a play with VideoServer2.py in the Samples folder of the NightDriverStrip project. It also sends video to Mesmeriser and Spectrum.

**Comment by robertlipe** [comment]
Additionally, the IP addresses  in them are hardcoded and not clearly
marked as examples. So be careful to dodge that, Harrie.

**Reply by heidepiek** [reply]
Trying, but til now no succes. 

![Clipboard01](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/104028471/90d1cc71-304b-41c8-a7ff-d818717e592e)

 What am i forgetting? I never worked with this software. I know i have uploaded the code right. My router gives the nightdriverstrip an ip adres:  
![Clipboard02](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/104028471/c31e8821-63f7-4d0e-a148-98e426aba06f)

found this: HostName: 192.168.8.198 

![Clipboard03](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/104028471/270760bc-9008-4ae8-b23a-7cf76b3a15c4)
 now connected i think, but i don`t see anything on the mesmerizer

 #ifndef PROJECT_NAME
    #define PROJECT_NAME                "Mesmerizer"
    #endif

    #define SHOW_FPS_ON_MATRIX          0
    #define ENABLE_AUDIOSERIAL          0   // Report peaks at 2400baud on serial port for PETRock consumption
    #define ENABLE_WIFI                 1   // Connect to WiFi
    #define INCOMING_WIFI_ENABLED       1   // Accepting incoming color data and commands
    #define WAIT_FOR_WIFI               1   // Hold in setup until we have WiFi - for strips without effects
    #define TIME_BEFORE_LOCAL           2   // How many seconds before the lamp times out and shows local content
    #define ENABLE_WEBSERVER            1  // Turn on the internal webserver
    #define ENABLE_NTP                  1   // Set the clock from the web
    #define ENABLE_OTA                  0   // Accept over the air flash updates
    #define ENABLE_REMOTE               1   // IR Remote Control
    #define ENABLE_AUDIO                1   // Listen for audio from the microphone and process it
    #define SCALE_AUDIO_EXPONENTIAL     0
    #define ENABLE_AUDIO_SMOOTHING      1
    #define EFFECT_PERSISTENCE_CRITICAL 1   // Require effects serialization to succeed

    #define INCOMING_CORE           0
    #define MAX_BUFFERS            40

**Reply by mikejohnau** [reply]
Yep, in Unity drop that IP address into the hostname field.

**Reply by heidepiek** [reply]
No connection, this is way out of my league! I hope someone can make a little tutorial. Maybe i`m using the wrong version of unity.
Some wrong settings? I don`t know

![no connection](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/104028471/9c68f6c6-68d4-46ac-8b20-5bc8fe90cc24)

**Reply by mikejohnau** [reply]
Try running VideoServer2.py you will need Python installed if you dont already have it. You will need to open the file and change the IP address. Let see if we can get some video showing at all.. Be prepared you may be Rickrolled. I cant remember with youtube video Dave used by default but one of them was Rick Astley.....

**Reply by heidepiek** [reply]
i give up. I think i`m doing it all wrong. Can`t find videoserver2.py. I think i`m using the wrong version of unity. I got it from the official site, but if i look at dave`s repository they talk about gist and other repository`s. It`s beyond my capabilities. In the unity nightdriver folder can`t find videoserver2.

**Reply by mikejohnau** [reply]
VideoServer2.py is in the NightDriverStrip/samples/videoserver folder.  It is seperate from Unity. I'm trying to work whether the issue is to do with Unity or something else.

**Reply by mikejohnau** [reply]
So on your PC, I dont know what OS you are running... but go to the NightDriverStrip/samples directory.

Edit the VideoServer2.py file and change the ESP32_WIFI_ADDRESS to the address of your ESP32 board.
the change the  settings below to match your matrix....
MATRIX_WIDTH = 64
MATRIX_HEIGHT = 32

Then if you have Python installed run "python3 VideoServer2.py"
 If you dont have Python installed.. Google how to install it for your OS.. and then run 'python3 VideoServer2.py'

You should get video on your matrix.

**Reply by heidepiek** [reply]
I have a fireplace now. But i`m a little flabbergasted. I have python 3 installed with the installer for windows. If i use command  C:\Users\Eigenaar>python -V
Python 3.12.0

it says version 3.12.0 But if i use your   "python3 VideoServer2.py" command i get this?

C:\Users\Eigenaar\Desktop\new icons nightdriver\NightDriverStrip-main(2)\NightDriverStrip-main\samples\videoserver>python3 VideoServer2.py
Python is niet gevonden. Voer zonder argumenten uit om te installeren vanuit de Microsoft Store of schakel deze snelkoppeling uit via Instellingen > Aliassen voor app-uitvoering beheren.

C:\Users\Eigenaar\Desktop\new icons nightdriver\NightDriverStrip-main(2)\NightDriverStrip-main\samples\videoserver>

But if i use: python VideoServer2.py    it works.

I have python 3 or don`t i ???

**Reply by mikejohnau** [reply]
ok. Lets just stop and celebrate that VideoServer2.py works.. On my count you are the 3rd person on the face of he planet to get this working..... Big achievement...  Go out have a few beers..

Yes you have Python..

But what we have proved is that you have enabled the incoming Wifi.

**Reply by mikejohnau** [reply]
I'm not sure on the python and python3... mine needs python3... 

Let look at Unity..
Can you edit  the LED Sign Controller script? either in Unity or outside of Unity??

**Reply by mikejohnau** [reply]
return LEDInterop.CombineByteArrays(LEDInterop.WORDToBytes(WIFI_COMMAND_PIXELDATA64),      // Offset, always zero for us
                                                    LEDInterop.WORDToBytes((UInt16)Channel),               // LED channel on ESP32
                                                    LEDInterop.DWORDToBytes((UInt32)data.Length /3),      // Number of LEDs
                                                    LEDInterop.ULONGToBytes(seconds),                      // Timestamp seconds (64 bit)
                                                    LEDInterop.ULONGToBytes(uSeconds),                     // Timestmap microseconds (64 bit)
                                                    data);  

This call haapens twice towards the end of the script. But in early versions the (Uint16) or (Unit32) are missing.. They are required..

**Reply by heidepiek** [reply]
i have opened  this file: LEDSignController.cs.

//                Debug.Log("Old Seconds: " + seconds + "uSec: " + uSeconds);

                return LEDInterop.CombineByteArrays(LEDInterop.WORDToBytes(WIFI_COMMAND_PIXELDATA),        // Offset, always zero for us
                                                    LEDInterop.WORDToBytes((UInt16)Channel),               // LED channel on ESP32
                                                    LEDInterop.DWORDToBytes((UInt32)data.Length / 3),      // Number of LEDs
                                                    LEDInterop.DWORDToBytes((UInt32)seconds),              // Timestamp seconds (32 bit truncation)
                                                    LEDInterop.DWORDToBytes((UInt32)uSeconds),             // Timestmap microseconds (32 bit truncation)
                                                    data);                                                 // Color Data
            }
            else
            {
            // The timeOffset is how far in the future frames are generated for.  If the chips have a 2 second buffer, you could
            // go up to 2 seconds, but I shoot for the middle of the buffer depth.  Right now it's calculated as using 

            double epoch = (timeStart.Ticks - 621355968000000000 + (1.0 * TimeSpan.TicksPerSecond)) / (double)TimeSpan.TicksPerSecond;
            double fraction = epoch - (Int64)epoch;

            ulong seconds = (ulong)epoch;                                       // Whole part of time number (left of the decimal point)
            ulong uSeconds = (ulong)(fraction * 1000000);           // Fractional part of time (right of the decimal point)

            // Debug.Log("New Seconds: " + seconds + "uSec: " + uSeconds);

                var data = GetPixelData(MainLEDs);
                return LEDInterop.CombineByteArrays(LEDInterop.WORDToBytes(WIFI_COMMAND_PIXELDATA64),      // Offset, always zero for us
                                                    LEDInterop.WORDToBytes((UInt16)Channel),               // LED channel on ESP32
                                                    LEDInterop.DWORDToBytes((UInt32)data.Length / 3),      // Number of LEDs
                                                    LEDInterop.ULONGToBytes(seconds),                      // Timestamp seconds (64 bit)
                                                    LEDInterop.ULONGToBytes(uSeconds),                     // Timestmap microseconds (64 bit)
                                                    data);                                                 // Color Data

Which version of unity are you using mike?

**Reply by mikejohnau** [reply]
2021.3.3211 Personal.. but I'm on MacOS..

**Reply by mikejohnau** [reply]
So let's clarify. VideoServer2.py gives you video  on your build?
But Unity with the same settings, IP address, matrix height, matrix width.... doesn't work? lets not talk quality...does stuff change  the matrix

**Reply by heidepiek** [reply]
Yes mike, videoserver2.py works: 

https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/104028471/eb126b25-40b0-479b-a6db-35065856f8c9

Unity i can`t get to work. I think a little incompetence on my part. Don't know exactly what to do with unity.

**Reply by heidepiek** [reply]
Thanks for your help mike, but for now i`m taking a break for this project. 👊

Greetz harrie

**Reply by mikejohnau** [reply]
I have just remembered that you also need to add 
 ${unity_double_flags.build_flags}

to your environment build flags in platform.io

**Comment by mikejohnau** [comment]
So we can forget about the pink pacman....... Changed panels and all is good now. It looks like my original 64x32 matrix has developed a fault.

**Reply by heidepiek** [reply]
Does happen that the rgb is not connected in the right order in the hardware of the matrix. Also have a few matrices lying around that are not connected properly. You can use them but you have to change the order of the r g b in the smartmatrix library. I already had that once with the aurora project on a 64x64 matrix. But if you want to use two panels, thats weird looking.

This is aurora by jason coon on a 64x64 matrix
https://drive.google.com/file/d/1wpPwLRDRy_e-BUn-FX8GWZmOtIP_DyUA/view?usp=sharing
